accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,718 @@
+  public sealed class RequiresDynamicCodeAttribute : CapabilityAttribute
+  {
+      public RequiresDynamicCodeAttribute();
+  }
+  ```
+
+  We instead suggest placing the equivalent `FeatureSwitch` attribute on the property definition:
+
+  ```csharp
+  public sealed class RuntimeFeature
+  {
+      [FeatureSwitch(typeof(RequiresDynamicCodeAttribute))]
+      public bool IsDynamicCodeSupported => // ...
+  }
+  ```
+
+  Both models allow multiple "feature switch" properties to be defined by the same feature, but the latter can prevent the same property from being marked as the "feature check" for multiple different features that could potentially conflict, if we set `AllowMultiple = false` on `FeatureSwitchAttribute`.
+
+- API shape: target of `CapabilityGuard` vs `FeatureGuard`
+
+  These are the same idea, and are represented very similarly in both models, targeting the guard property, with `AllowMultiple = true` to support a property that guards multiple features. The capability API draft lists the attribute targets as `AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Field`. We might want to start with just `AttributeTargets.Property`.
+
+## Comparison with platform compatibility analyzer
+
+The platform compatibility analyzer is semantically very similar to the behavior described here, except that it doesn't come with ILLink/ILCompiler support for removing branches that are unreachable when publishing for a given platform.
+
+"Platforms" (instead of "features") are represented as strings, with optional versions.
+
+- `SupportedOSPlatformAttribute` is similar to the `RequiresDynamicCodeAttribute`, etc, and will produce warnings if called in an unguarded context.
+
+  ```csharp
+  CallSomeAndroidAPI(); // warns if not targeting android
+
+  [SupportedOSPlatform("android")]
+  static void CallSomeAndroidAPI() {
+    // ...
+  }
+  ```
+
+- Platform checks are like feature switch properties, and can guard calls to annotated APIs:
+
+  ```csharp
+  if (OperatingSystem.IsAndroid())
+    CallSomeAndroidAPI(); // no warning for guarded call
+  ```
+
+  The analyzer has built-in knowledge of fact that `IsAndroid` corresponds to the `SupportedOSPlatform("android")`. This is similar to the ILLink analyzer's current hard-coded knowledge of the fact that `IsDynamicCodeSupported` corresponds to `RequiresDynamicCodeAttribute`.
+
+- Platform guards are like feature guards, allowing libraries to introduce custom guards for existing platforms:
+
+  ```csharp
+  class Feature {
+      [SupportedOSPlatformGuard("android")]
+      public static bool IsSupported => SomeCondition() && OperatingSystem.IsAndroid();
+  }
+  ```
+
+  ```csharp
+  if (Feature.IsSupported)
+    CallSomeAndroidAPI(); // no warning for guarded call
+  ```
+
+The platform compatibility analyzer also has some additional functionality, such as annotating _unsupported_ APIs, and including version numbers.
+
+## Possible future extensions
```

Comment:
One other future thing to look at would be the getter body for a property with a `FeatureSwitch` attribute. Currently these should follow a set of rules for the system to work correctly. Having either a way to validate it, or to generate these would help with consistency and robustness. These new attributes should provide enough information for this to be possible.